lightning_examples/finetuning-scheduler/finetuning-scheduler.py

```diff
@@ -0,0 +1,612 @@
+# large transformer-based language models. The values used here have some justification
+# in the referenced literature but have been largely empirically determined and while a good
+# starting point could be could be further tuned.
+#
+# </div>
+
+# %%
+optimizer_init = {
+    "class_path": "torch.optim.AdamW",
+    "init_args": {"weight_decay": 1e-05, "eps": 1e-07, "lr": 1e-05},
+}
+
+# %% [markdown]
+# ### LR Scheduler Configuration
+#
+# <div id="a3">
+#
+# The [CosineAnnealingWarmRestarts scheduler](https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.CosineAnnealingWarmRestarts.html?highlight=cosineannealingwarm#torch.optim.lr_scheduler.CosineAnnealingWarmRestarts) nicely fits with our iterative finetuning since it does not depend upon a global max_epoch
+# value. The importance of initial warmup is reduced due to the innate warmup effect of Adam bias correction [[5]](#f3)
+# and the gradual thawing we are performing. Note that commonly used LR schedulers that depend on providing
+# max_iterations/epochs (e.g. the
+# [CosineWarmupScheduler](https://github.com/PyTorchLightning/lightning-tutorials/blob/0c325829101d5a6ebf32ed99bbf5b09badf04a59/course_UvA-DL/05-transformers-and-MH-attention/Transformers_MHAttention.py#L688)
+# used in other pytorch-lightning tutorials) also work with FinetuningScheduler. Though the LR scheduler is theoretically
+# justified [(Loshchilov & Hutter, 2016)](#f4), the particular values provided here are primarily empircally driven.
+#
+# </div>
+
+
+# %%
+lr_scheduler_init = {
+    "class_path": "torch.optim.lr_scheduler.CosineAnnealingWarmRestarts",
+    "init_args": {"T_0": 1, "T_mult": 2, "eta_min": 1e-07},
+}
+pl_lrs_cfg = {"interval": "epoch", "frequency": 1, "name": "CosineAnnealingWarmRestarts"}
+
+# %%
+model = RteBoolqModule(
+    model_name_or_path="albert-base-v2",
+    optimizer_init=optimizer_init,
+    lr_scheduler_init=lr_scheduler_init,
+    pl_lrs_cfg=pl_lrs_cfg,
+    experiment_tag="fts_explicit",
+)
+callbacks = [
+    FinetuningScheduler(ft_schedule=ft_schedule_name, max_depth=2),  # type: ignore # noqa
+    FTSEarlyStopping(monitor="val_loss", min_delta=0.001, patience=2),  # type: ignore # noqa
+    FTSCheckpoint(monitor="val_loss", save_top_k=5),  # type: ignore # noqa
+]
+example_logdir = "lightning_logs"
+logger = TensorBoardLogger(example_logdir, name="fts_explicit")
+
+# %%
+# optionally start tensorboard and monitor progress graphically while viewing multi-phase finetuning specific training
+# logs in the cell output below by uncommenting the next 3 lines
+# # !mkdir -p $example_logdir
+# # %load_ext tensorboard
+# # %tensorboard --logdir example_logdir
+# disable progress bar by default to focus on multi-phase training logs. Set to True to re-enable if desired
+enable_progress_bar = False
+
+# %%
+def train() -> None:
+    trainer = pl.Trainer(
+        enable_progress_bar=enable_progress_bar,
+        precision=16,
+        gpus=1,
+        # accelerator="auto",
```

Comment:
Should we uncomment this ? If you want to stick with gpus=1, let's use devices=1, accelerator="gpu"

Reply:
Done!